cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.queue;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.module.EmbeddedDataset;
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.TriggerJsonCodec;
+import co.cask.cdap.internal.schedule.trigger.Trigger;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.id.ScheduleId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Dataset that stores {@link Job}s, which correspond to schedules that have been triggered, but not yet executed.
+ *
+ * Row Key is in the following formats:
+ *   For Jobs:
+ *     'J':<partition_id>:<scheduleId>:<timestamp>
+ *     The <partition_id> is a hash based upon the scheduleId
+ *
+ *   For TMS MessageId:
+ *     'M':<topic>
+ */
+public class JobQueueDataset extends AbstractDataset implements JobQueue {
+
+  static final String EMBEDDED_TABLE_NAME = "t"; // table
+  private static final Gson GSON =
+    new GsonBuilder().registerTypeAdapter(Trigger.class, new TriggerJsonCodec()).create();
+
+  // For now, simply serialize the entire Job into one column
```

Comment:
for now? Is there an improvement needed? In that case please mark as TODO and make sure you address it in 4.2. To avoid future need for migration